src/common/KokkosKernels_ExecSpaceUtils.hpp

```diff
@@ -136,7 +137,271 @@ inline int kk_get_suggested_team_size(const int vector_size, const ExecSpaceType
+  }
+
+  template<typename Tag = void>
+  Kokkos::TeamPolicy<ExecSpace, Tag> team_policy(size_t num_teams, size_t team_size)
+  {
+    if(!recording)
+      throw std::runtime_error("CudaGraphWrapper: Can't create TeamPolicy because begin_recording() has not been called");
+    return Kokkos::TeamPolicy<ExecSpace, Tag>(num_teams, team_size);
+  }
+
+  template<typename Tag = void>
+  Kokkos::TeamPolicy<ExecSpace, Tag> team_policy(size_t num_teams, size_t team_size, size_t vector_size)
+  {
+    if(!recording)
+      throw std::runtime_error("CudaGraphWrapper: Can't create TeamPolicy because begin_recording() has not been called");
+    return Kokkos::TeamPolicy<ExecSpace, Tag>(num_teams, team_size, vector_size);
+  }
+
+  template<typename Tag = void>
+  Kokkos::TeamPolicy<ExecSpace, Tag> team_policy(size_t num_teams, size_t team_size, size_t vector_size, size_t sharedPerTeam, size_t sharedPerThread)
+  {
+    if(!recording)
+      throw std::runtime_error("CudaGraphWrapper: Can't create TeamPolicy because begin_recording() has not been called");
+    return Kokkos::TeamPolicy<ExecSpace, Tag>(num_teams, team_size).set_scratch_size(0, Kokkos::PerTeam(sharedPerTeam), Kokkos::PerThread(sharedPerThread));
+  }
+
+  template<typename Tag = void>
+  Kokkos::RangePolicy<ExecSpace, Tag> range_policy(size_t begin, size_t end)
+  {
+    if(!recording)
+      throw std::runtime_error("CudaGraphWrapper: Can't create RangePolicy because begin_recording() has not been called");
+    return Kokkos::RangePolicy<ExecSpace, Tag>(begin, end);
+  }
+
+  void launch() {
+    // The kernels were executed during recording, but fence now
+    ExecSpace().fence();
+  }
+
+  bool recording = false;
+};
+
+#if defined(KOKKOS_ENABLE_CUDA) && 10010 <= CUDA_VERSION
+#define HAVE_CUDAGRAPHS
+
+//Dummy parallel call that makes compiler generate some declarations/initializations
+//necessary for CUDA graph stream capture to work (see Kokkos issue #2606)
+namespace Dummy
+{
+  struct F
+  {
+    KOKKOS_INLINE_FUNCTION void operator()(const int) const {}
+  };
+  inline void f()
+  {
+    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::Cuda>(0,1), F());
+  }
+}
+
+//Specialize the wrapper for Cuda, to actually use the graph features.
+template<typename LaunchParams>
+struct CudaGraphWrapper<Kokkos::Cuda, LaunchParams>
+{
+  //The class assumes these are 
+  cudaGraph_t graph;
+  cudaGraphExec_t instance;
+  cudaStream_t stream;
+  Kokkos::Cuda kokkosStream;
+
+  using P_t = typename Kokkos::Experimental::WorkItemProperty::HintLightWeight_t;
+  static constexpr P_t P = Kokkos::Experimental::WorkItemProperty::HintLightWeight;
+  
+  using WrappedTeamPolicy =
+    typename Kokkos::Experimental::Impl::PolicyPropertyAdaptor<
+    P_t, Kokkos::TeamPolicy<Kokkos::Cuda>>::policy_out_t;
+  using WrappedRangePolicy =
+    typename Kokkos::Experimental::Impl::PolicyPropertyAdaptor<
+    P_t, Kokkos::RangePolicy<Kokkos::Cuda>>::policy_out_t;
+
+  #define CUDA_CALL(x) \
```

Comment:
This is a popular macro to define, so this definition will quite likely collide with other code.  In general, it's unfortunate to expose CUDA API functions in a header file.

Reply:
@mhoemmen Good points. None of the native CUDA API calls actually depend on template parameters so I will put those in a separate .cpp.

Reply:
Actually this already exists in Kokkos, if you grep for `CUDA_SAFE_CALL` you will find an implementation of this.
Why not use the macro from Kokkos, this would avoid code duplication...

Reply:
@lucbv Kokkos itself really shouldn't be defining `CUDA_SAFE_CALL`, since it's not namespaced.  It's also a common macro that CUDA documentation seems to ask users to define.